### PR TITLE
Add tool to extract a unique list of expression symbols.

### PR DESCRIPTION
**Pull request by samueltootle**
Generate a unique list of expression symbols with optional filtering of known symbols, e.g. coordinates.

### CODE REVIEW

**Review by zachetienne — COMMENTED** [review]
Sympy's expr.free_symbols already outputs the unique list of symbols for an expression.

**Comment by samueltootle** [comment]
Thank you for pointing this out.  I have replaced the internal algorithm to use sympy as this is much cleaner.  The utility function is still since this is needed in multiple places in an upcoming PR

File: nrpy/helpers/expr_tree.py [hunk 1]
```diff
@@ -190,6 +190,31 @@ def __repr__(self) -> str:
...
+
+    :param expr: Sympy expression
+    :param exclude: List of symbol names to exclude
+    :returns: List of unique symbols from the expression
+
+    DOCTEST:
+    >>> from sympy.abc import a, b
+    >>> from sympy import cos, sin
+    >>> xx0 = sp.Symbol('xx0')
+    >>> x = cos(a + b)**2 ++ xx0
```
Comment (zachetienne) [inline nrpy/helpers/expr_tree.py:207]: Bug in this line of code: ++

File: nrpy/helpers/expr_tree.py [hunk 2]
```diff
@@ -190,6 +190,31 @@ def __repr__(self) -> str:
...
+    >>> get_unique_expression_symbols(x, ["xx0"])
```
Comment (zachetienne) [inline nrpy/helpers/expr_tree.py:208]: Bug in this line of code: "xx0" is a string.

File: nrpy/helpers/expr_tree.py [hunk 3]
```diff
@@ -190,6 +190,31 @@ def __repr__(self) -> str:
...
+    """
+    if exclude is None:
+        exclude = []
+
+    symbols = {str(sym) for sym in expr.free_symbols}
+    return sorted([sym for sym in symbols if not sym in exclude])
+
+
 if __name__ == "__main__":
     import doctest
```
Comment (zachetienne) [inline nrpy/helpers/expr_tree.py:219]: Missing required code for executing doctests:
```
if __name__ == "__main__":
    import doctest

    results = doctest.testmod()

    if results.failed > 0:
        print(f"Doctest failed: {results.failed} of {results.attempted} test(s)")
        sys.exit(1)
    else:
        print(f"Doctest passed: All {results.attempted} test(s) passed")
```

File: nrpy/helpers/expr_tree.py [hunk 4]
```diff
@@ -190,6 +190,31 @@ def __repr__(self) -> str:
...
+    >>> xx0 = sp.Symbol('xx0')
+    >>> x = cos(a + b)**2 ++ xx0
+    >>> get_unique_expression_symbols(x, ["xx0"])
+    ['a', 'b']
+    """
+    if exclude is None:
+        exclude = []
+
+    symbols = {str(sym) for sym in expr.free_symbols}
+    return sorted([sym for sym in symbols if not sym in exclude])
```
Comment (zachetienne) [inline nrpy/helpers/expr_tree.py:215]: I don't like that we're converting the symbols to strings here, as it's a recipe for nonrobustness (and slowness too!) Why can't exclude be a list of symbols?

Comment (samueltootle) [reply nrpy/helpers/expr_tree.py:215]: In many places, the inclusion of `xx0`, `xx1`, `xx2` is hard coded in function bodies, thus the current use, e.g., is the following:
`unique_symbols += get_unique_expression_symbols(expr, exclude=[f"xx{i}" for i in range(3)])`

Unfortunately this cannot be changed to something like:
`unique_symbols += get_unique_expression_symbols(expr, exclude=[sp.Symbol(f"xx{i}") for i in range(3)])`

as these are unique Sympy objects and thus cannot be compared.

Comment (zachetienne) [reply nrpy/helpers/expr_tree.py:215]: Actually, I think the latter (comparing symbols) *will* work! You may need to ensure that real=True is set properly.

Comment (samueltootle) [reply nrpy/helpers/expr_tree.py]: `expr_tree.py` is no longer modified.  I have moved this functionality `nrpy/helpers/expression_utils.py` per our slack discussion after updating the internal algorithm and function name.

**Pull request merged**